### ISSUE TITLE
High latency in TF propagation 

### ISSUE THREAD

**Issue by Myzhar** [ros2/rviz#359]
I'm porting the Stereolabs ZED 3D camera package to ROS2 and I cannot work with TF because of the high latency.
I'm sure that the TF messages propagates because I can see them using the `ros2 topic` and the `ros2 run tf2_ros tf2_echo` CLI commands... but Rviz receives them with many seconds of latency (sometimes I see frame axis move in Rviz after 2 or 3 three seconds after I stopped the node).

I thought that it was a problem with our node until I made a test with the "dummy robot" example and I figured out that the latency is a general problem:

 * Run the example:
   `$ ros2 launch dummy_robot_bringup dummy_robot_bringup.launch.py`
* Open `$ rviz2`
* Set the fixed frame to `world`
* Add a TF plugin viewer. You will see the robot link moving
* Add a `LaserScan` plugin and subscribe to `/scan`

At this point you will see the simulated laser scan on the oscillating link of the robot and you will notice how it is rendered in late respect to its reference frame.
Furthermore you will notice a lot of errors 
`Could not transform from [single_rrbot_hokuyo_link] to [world]` in the `LaserScan` plugin
and 
`[ERROR] [rviz2]: Lookup would require extrapolation into the future.  Requested time 1538495387.01342 but the latest data is at time 1538495386.98954, when looking up transform from frame [single_rrbot_hokuyo_link] to frame [world]`
in the console where RVIZ2 has been executed

![image](https://user-images.githubusercontent.com/3648617/46360591-0fe16f00-c66c-11e8-82e2-162bd7146734.png)

I read that a solution was provided for a similar issue and I updated ROS2 and Rviz2 to the latest version of the `master` branch, but the issue is still present.
I also tried to change RMW, but each of the three available options have the same problem.

**Comment by Martin-Idel-SI** [comment]
This is connected to https://github.com/ros2/rviz/issues/332. 
As far as I can see the problem is the following: the laser scan display currently asks for a transform at specific times - for a laser scan these times are not just one time point, but a range of times. Previously, the message filter made sure that the transform was available by allowing for some tolerance and doing some stuff behind the scene, but since message filters weren't available in ros2 until very recently, they aren't used, yet. 

I would expect this bug to persist until https://github.com/ros2/rviz/issues/55 is resolved.

**Comment by Myzhar** [comment]
I read the #332, but it does not explain why I get the TF updates in Rviz2 even after that the node has been stopped. It seems like there is somewhere a queue that is processed very slowly

**Comment by Martin-Idel-SI** [comment]
Hmm..., I can't reproduce this. 

I see the following:
- the laser scan is only very rarely rendered due to missing transforms and therefore it often seems as if it was late but that's only because I don't see the latest scan but an old one (at least in your image, the rendered scan is also not the last scan message you received as that one wasn't rendered due to missing transform). 
- When I stop the dummy robot, no new laser scan gets rendered at the same moment that no new tf2 information gets processed. Of course, the last rendered laser scan will rarely be at the head of the robots last position, because it gets rendered only very rarely. 
- In other words: I see the same as your screenshot, but I can also see that the laser scan rendered on screen is maybe a second old, but not because of latency but because all updates between then and now failed due to missing transforms.

These issues are related to what type of tf2 information `laser_geometry` requires and should be fixed by using message filters (might be a second issue, too, but without message filters, it's hard to see).

**Comment by Myzhar** [comment]
OK, so let's wait for message filters re introduction and let's see if it solves the problem.

**Comment by Martin-Idel-SI** [comment]
Yes, I believe this will make things much more apparent. But thanks for reporting the issue - maybe it turns out there is another problem.

**Comment by wjwwood** [comment]
We're waiting on tf message filter.

**Comment by mjcarroll** [comment]
I've started looking at tf message filter, will update with a PR reference
when I've got something.

**Comment by Squadrick** [comment]
@wjwwood Now that tf2_message filter has been merged with this [PR](https://github.com/ros2/geometry2/pull/81), when will rviz be updated?

**Comment by wjwwood** [comment]
I don't know, I don't have time to do it right now, but if someone did a pr, I might have time to review it. Can't promise when though.

**Comment by jacobperron** [comment]
@Myzhar Since #375 has been merged, can you check if you still have this issue?

**Comment by Myzhar** [comment]
@jacobperron I'm currently on another project and I don't know when I have free time to compile ROS2 from source.
I hope I can test it very soon

**Comment by allenh1** [comment]
@Myzhar I build a daily docker container with ROS 2 from the current master that you could probably use for your tests, should you be interested. It's `allenh1/rplidar_ros_demo:latest`.

**Comment by ymd-stella** [comment]
I did some research on this issue. 
The summary and my opinions are summarized below.

#### Summary
* qos of tf2_ros::TransformListener is `DynamicListenerQoS(depth=100)`.
  * [tf2_ros/include/tf2_ros/qos.hpp#L42](https://github.com/ros2/geometry2/blob/2e4c23caf9f17104e42976c5f8236a34496e03f0/tf2_ros/include/tf2_ros/qos.hpp#L42)
  * [tf2_ros/include/tf2_ros/transform_listener.h#L58-L62](https://github.com/ros2/geometry2/blob/a0cd9d73241edd25b94aa13a535c1bd2e6be1aeb/tf2_ros/include/tf2_ros/transform_listener.h#L58-L62)
  * https://github.com/ros2/rviz/blob/c5e68039ea5329f6e15216949c8429fb5ad21e2f/rviz_default_plugins/src/rviz_default_plugins/transformation/tf_wrapper.cpp#L126-L127
* `spin_some(10ms)` is processed in the same loop as the rendering process.
https://github.com/ros2/rviz/blob/c5e68039ea5329f6e15216949c8429fb5ad21e2f/rviz_common/src/rviz_common/visualization_manager.cpp#L390
* `spin_some(10ms)` is called at 30 fps.
* It seems that `spin_some(10ms)` reads only one message of `/tf` from the message queue.

#### The principle of the problem
* If the output frequency of `/tf` exceeds 30 Hz, the message will accumulate up to `depth(=100)` in the subscription queue of `/tf`.
* The maximum delay is 100 times the output period of `/tf`.

#### About solutions
* Execute tf2_ros::TransformListener in the other thread.
  * In rviz, tf2_ros::TransformListener's spin_thread was set to true, so this problem did not occur.
  * relevant PR is https://github.com/ros2/rviz/pull/346 and https://github.com/ros2/rviz/pull/340
  * relevant commit is https://github.com/bosch-io/rviz/commit/b27d8f6335d92b14e5047ed777dd2d2c3e7714fd#diff-a9019488dd7d224ab0d14c8e962c4209R142
